### PR TITLE
[libmicrohttpd] Add support for Windows ARM builds

### PR DESCRIPTION
**Pull request by anmatako**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] ~SHA512s are updated for each updated download~
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] ~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~
- [x] ~Any patches that are no longer applied are deleted from the port's directory.~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/libmicrohttpd/0001-Add-ARM-configurations.patch [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
++#if defined(_M_X64) || defined(_M_AMD64) || defined(_M_ARM64)
+ #define W32_VARDECORPREFIX
+ #define W32_DECORVARNAME(v) v
+ #define W32_VARDECORPEFIXSTR ""
+-#elif defined(_M_IX86) || defined(_X86_)
++#elif defined(_M_IX86) || defined(_X86_) || defined(_M_ARM)
+ #define W32_VARDECORPREFIX _
+ #define W32_DECORVARNAME(v) _ ## v
+ #define W32_VARDECORPEFIXSTR "_"
+diff --git a/w32/VS2015/hellobrowser.vcxproj b/w32/VS2015/hellobrowser.vcxproj
```
Comment (anmatako) [inline ports/libmicrohttpd/0001-Add-ARM-configurations.patch:32]: This patch is minimally modifying only the VS2015 solution that actually gets built by the port. The other flavors for VS2017, 2019, and 2022 can be updated similarly if there's desire to build one of these instead.

**Comment by BillyONeal** [comment]
Has this change been submitted upstream?

**Comment by anmatako** [comment]
> Has this change been submitted upstream?

I'm not sure I understand the question. It's my first time opening a PR for vcpkg. I though that submitting the PR to the repo is what I was supposed to do for the upstream.

**Comment by BillyONeal** [comment]
> I'm not sure I understand the question. It's my first time opening a PR for vcpkg. I though that submitting the PR to the repo is what I was supposed to do for the upstream.

Ah, sorry for the confusion. This seems like a bugfix to the build system of libmicrohttpd; there's no reason that the fix should be specific to vcpkg. In general we like to make sure upstream, in this case the folks who make libmicrohttpd, are at least notified when there are not specific-to-vcpkg patches like this. It helps avoid situations like https://wiki.debian.org/SSLkeys#End_User_Summary where Debian added OpenSSL patches that broke key generation.

Because this seems to only touch build system stuff we don't necessarily need to wait for those folks to have merged the changes before proceeding, but we would prefer if they were at least notified.

**Comment by anmatako** [comment]
> > I'm not sure I understand the question. It's my first time opening a PR for vcpkg. I though that submitting the PR to the repo is what I was supposed to do for the upstream.
> 
> Ah, sorry for the confusion. This seems like a bugfix to the build system of libmicrohttpd; there's no reason that the fix should be specific to vcpkg. In general we like to make sure upstream, in this case the folks who make libmicrohttpd, are at least notified when there are not specific-to-vcpkg patches like this. It helps avoid situations like https://wiki.debian.org/SSLkeys#End_User_Summary where Debian added OpenSSL patches that broke key generation.
> 
> Because this seems to only touch build system stuff we don't necessarily need to wait for those folks to have merged the changes before proceeding, but we would prefer if they were at least notified.

OK I see what you mean now :). Given how this library is maintained it's hard to figure out if there's even a plan to support ARM builds. I don't see any open issues regarding that or a way to submit a PR for their build system. I understand if you're unwilling to take this change in VCPKG and I can find an alternative path forward for the use-case I'm trying to cover on my end,

**Comment by anmatako** [comment]
@BillyONeal I will notify the microhttpd mailing list regarding the ARM support and the related change posted here. Seeing that a change from 0.9.75 to 0.9.76 took over a year, a response and/or acknowledgement may take a while :)

**Comment by BillyONeal** [comment]
> @BillyONeal I will notify the microhttpd mailing list regarding the ARM support and the related change posted here. Seeing that a change from 0.9.75 to 0.9.76 took over a year, a response and/or acknowledgement may take a while :)

Sure, like I said, we don't absolutely need them to have responded with a release or similar, but we do want to see that they at least have been notified and haven't explicitly said 'no we don't want that' or similar.

> The proper fix will be provided by the upstream. You can close this PR.

Is it expected that they'll drop a release for that or should we just add whatever patch they decided to do?

**Comment by anmatako** [comment]
@BillyONeal I just reached out to the mailing list yesterday and have not heard back yet. I'll wait to see if there is a response on their end with willingness (or unwillingness) to take the change on their end. In the case the upstream is not interested in supporting an ARM build, will this be acceptable to take this change into VCPKG or will the path forward be blocked?

**Comment by BillyONeal** [comment]
> @BillyONeal I just reached out to the mailing list yesterday and have not heard back yet. I'll wait to see if there is a response on their end with willingness (or unwillingness) to take the change on their end. In the case the upstream is not interested in supporting an ARM build, will this be acceptable to take this change into VCPKG or will the path forward be blocked?

I think we should give them 30 days to respond but if they don't we'll take the patch. Is there a link to a mailing list archive or something we can record here?

**Comment by anmatako** [comment]
> > @BillyONeal I just reached out to the mailing list yesterday and have not heard back yet. I'll wait to see if there is a response on their end with willingness (or unwillingness) to take the change on their end. In the case the upstream is not interested in supporting an ARM build, will this be acceptable to take this change into VCPKG or will the path forward be blocked?
> 
> I think we should give them 30 days to respond but if they don't we'll take the patch. Is there a link to a mailing list archive or something we can record here?

One thing I did not realize till later when I saw the mailing list digest is that @Karlson2k is involved in microhttpd and willing to take the upstream changes needed to enable the ARM builds. Thank you @Karlson2k for that!

@BillyONeal given that, I think I will leave this PR open until the upstream is updated and then I will update the port with the correct version bump for httpd with ARM support.

Does that sound reasonable?

BTW this is the mailing list archive for reference https://lists.gnu.org/archive/html/libmicrohttpd/2023-04/msg00003.html

**Comment by BillyONeal** [comment]
> I think I will leave this PR open until the upstream is updated

Sounds good, just un-draft the PR when you are ready by clicking the 'ready for review' button :)

**Comment by Karlson2k** [comment]
> Is it expected that they'll drop a release for that or should we just add whatever patch they decided to do?

I'm the upstream maintainer.
The new release will be made in foreseeable future.

The support will be implemented in git master soon.

You can wait for the new release (with other fixes) or just cherry-pick commits for ARM/ARM64 support in VS project files.

BTW I see other problems in current portfile: wrong set of headers used, completely disabled ARM and others. ARM is fully supported when libmicrohttpd built by GCC or clang. Only MSVC implementation is not complete.
Probably I'll open PR to fix these and other problems.

**Comment by Karlson2k** [comment]
The Visual Studio with ARM/ARM64 support has been added in git master. The diff:
* https://git.gnunet.org/libmicrohttpd.git/diff/?id2=55aa3677af6065c96130076c85e3c672e7ba0a49&id=9942b6415c4f1a50dba3306e4fc42cb709e41143
* https://github.com/Karlson2k/libmicrohttpd/compare/55aa3677af6065c96130076c85e3c672e7ba0a49...9942b6415c4f1a50dba3306e4fc42cb709e41143

After some testing it will be pulled to the next version branch and then published.

**Comment by anmatako** [comment]
> The Visual Studio with ARM/ARM64 support has been added in git master. The diff:
> 
> * https://git.gnunet.org/libmicrohttpd.git/diff/?id2=55aa3677af6065c96130076c85e3c672e7ba0a49&id=9942b6415c4f1a50dba3306e4fc42cb709e41143
> * [Karlson2k/libmicrohttpd@55aa367...9942b64](https://github.com/Karlson2k/libmicrohttpd/compare/55aa3677af6065c96130076c85e3c672e7ba0a49...9942b6415c4f1a50dba3306e4fc42cb709e41143)
> 
> After some testing it will be pulled to the next version branch and then published.

@Karlson2k Thank you so much for the upstream update. I'm trying it out now by taking your patch into my vcpkg fork so I can build it through the pipeline and test it out. It will take a day or two to find the time to test the ARM builds, but I will let you know as soon as this is done.

**Comment by anmatako** [comment]
> > The Visual Studio with ARM/ARM64 support has been added in git master. The diff:
> > 
> > * https://git.gnunet.org/libmicrohttpd.git/diff/?id2=55aa3677af6065c96130076c85e3c672e7ba0a49&id=9942b6415c4f1a50dba3306e4fc42cb709e41143
> > * [Karlson2k/libmicrohttpd@55aa367...9942b64](https://github.com/Karlson2k/libmicrohttpd/compare/55aa3677af6065c96130076c85e3c672e7ba0a49...9942b6415c4f1a50dba3306e4fc42cb709e41143)
> > 
> > After some testing it will be pulled to the next version branch and then published.
> 
> @Karlson2k Thank you so much for the upstream update. I'm trying it out now by taking your patch into my vcpkg fork so I can build it through the pipeline and test it out. It will take a day or two to find the time to test the ARM builds, but I will let you know as soon as this is done.

FYI @Karlson2k I was able to confirm the ARM binaries of libmicrohttpd through our internal testing. Let us know when you get the changes merged into the upstream and there's a new version out.

**Comment by Karlson2k** [comment]
> FYI @Karlson2k I was able to confirm the ARM binaries of libmicrohttpd through our internal testing. Let us know when you get the changes merged into the upstream and there's a new version out.
Thanks for the testing and for the confirmation, @anmatako.
The changes are already merged upstream. The new version should be released in one or two weeks.

**Comment by jimwang118** [comment]
@anmatako The upstream has incorporated the patch that supports `arm`, you can update the upstream version to complete this PR.

**Comment by Karlson2k** [comment]
Actually the fixed version was already merged via PR #31740 so probably this PR should be closed.

**Comment by jimwang118** [comment]
Duplicate of https://github.com/microsoft/vcpkg/pull/31740

**Pull request closed without merging**